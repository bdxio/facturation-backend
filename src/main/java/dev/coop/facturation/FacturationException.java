package dev.coop.facturation;

/**
 *
 * @author lfo
 */
public class FacturationException extends RuntimeException {

    public FacturationException() {
    }

    public FacturationException(String message) {
        super(message);
    }

    public FacturationException(String message, Throwable cause) {
        super(message, cause);
    }

    public FacturationException(Throwable cause) {
        super(cause);
    }
    
    
}
