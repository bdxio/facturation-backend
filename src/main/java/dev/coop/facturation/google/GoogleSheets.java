package dev.coop.facturation.google;

import com.google.api.client.googleapis.auth.oauth2.GoogleCredential;
import com.google.gdata.client.spreadsheet.SpreadsheetService;
import com.google.gdata.client.spreadsheet.WorksheetQuery;
import com.google.gdata.data.spreadsheet.WorksheetEntry;
import com.google.gdata.data.spreadsheet.WorksheetFeed;
import com.google.gdata.util.ServiceException;
import java.io.IOException;
import java.net.URL;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.util.Assert;

/**
 *
 * @author lfo
 */
@Service
public class GoogleSheets {

    private SpreadsheetService service;

    @Autowired
    public GoogleSheets(GoogleCredential googleCredential) {
        Assert.notNull(googleCredential);
        service = new SpreadsheetService("compta-devcoop");
        service.setProtocolVersion(SpreadsheetService.Versions.V3);
        service.setOAuth2Credentials(googleCredential);
    }

    public WorksheetEntry getWorksheetEntry(WorksheetFeed feed, String worksheetName) {
        return feed.getEntries()
                .stream()
                .filter(w -> w.getTitle().getPlainText().equals(worksheetName))
                .findFirst().get();
    }

    public WorksheetFeed getWorksheetFeed(String spreadsheetId) {
        try {
            String path = String.format("https://spreadsheets.google.com/feeds/worksheets/%s/private/full", spreadsheetId);
            WorksheetQuery query = new WorksheetQuery(new URL(path));
            WorksheetFeed feed = service.query(query, WorksheetFeed.class);
            return feed;
        } catch (IOException | ServiceException ex) {
            throw new GsException(ex);
        }
    }

    public SpreadsheetService getService() {
        return service;
    }

}
