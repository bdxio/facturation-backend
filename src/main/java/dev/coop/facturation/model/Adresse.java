package dev.coop.facturation.model;

public class Adresse {

    private String rue1;
    private String rue2;
    private String rue3;
    private String codePostal;
    private String ville;
    private String pays;

//    private Adresse(String rue1, String rue2, String rue3, String codePostal, String ville, String pays) {
//        this.rue1 = rue1;
//        this.rue2 = rue2;
//        this.rue3 = rue3;
//        this.codePostal = codePostal;
//        this.ville = ville;
//        this.pays = pays;
//    }
    public String getRue1() {
        return rue1;
    }

    public String getRue2() {
        return rue2;
    }

    public String getRue3() {
        return rue3;
    }

    public String getCodePostal() {
        return codePostal;
    }

    public String getVille() {
        return ville;
    }

    public String getPays() {
        return pays;
    }

    public Adresse setRue1(String r1) {
        rue1 = r1;
        return this;
    }

    public Adresse setRue2(String r2) {
        rue2 = r2;
        return this;
    }

    public Adresse setRue3(String r3) {
        rue3 = r3;
        return this;
    }

    public Adresse setCodePostal(String cP) {
        codePostal = cP;
        return this;
    }

    public Adresse setVille(String v) {
        ville = v;
        return this;
    }

    public Adresse setPays(String p) {
        pays = p;
        return this;
    }

}