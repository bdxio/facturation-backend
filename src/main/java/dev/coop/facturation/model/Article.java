package dev.coop.facturation.model;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility;

@JsonAutoDetect(fieldVisibility = Visibility.ANY, getterVisibility = Visibility.NONE, setterVisibility = Visibility.NONE)
public class Article extends HasSocieteCodeKey {

    private String description;
    private Montant montant;
    private Unite unite;
    private TVA tva;

    public Article(Societe societe, int codeValue) {
        super(societe, codeValue);
    }

    @Override
    public CodeFormatter.Prefix getCodePrefix() {
        return CodeFormatter.Prefix.AR;
    }

    public String getDescription() {
        return description;
    }

    public Article setDescription(String description) {
        this.description = description;
        return this;
    }

    public Montant getMontant() {
        return montant;
    }

    public Article setMontant(Montant montant) {
        this.montant = montant;
        return this;
    }

    public Unite getUnite() {
        return unite;
    }

    public Article setUnite(Unite unite) {
        this.unite = unite;
        return this;
    }

    public TVA getTva() {
        return tva;
    }

    public Article setTva(TVA tva) {
        this.tva = tva;
        return this;
    }

}
