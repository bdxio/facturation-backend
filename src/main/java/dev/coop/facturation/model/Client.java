package dev.coop.facturation.model;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import org.springframework.data.annotation.PersistenceConstructor;

/**
 *
 * @author lforet
 */
@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY, getterVisibility = JsonAutoDetect.Visibility.NONE, setterVisibility = JsonAutoDetect.Visibility.NONE)
public class Client extends HasSocieteCodeKey {
        
    private String nom;
    private String nomCourt;
    private String numTVAIntracom;
    
    private Adresse adresse;

    @PersistenceConstructor
    public Client(SocieteCodeKey id) {
        super(id);
    }

    public Client(Societe societe, int codeValue) {
        super(societe, codeValue);
    }
    
    @Override
    public CodeFormatter.Prefix getCodePrefix() {
        return CodeFormatter.Prefix.CL;
    }

    public String getNom() {
        return nom;
    }

    public Client setNom(String nom) {
        this.nom = nom;
        return this;
    }

    public Adresse getAdresse() {
        return adresse;
    }

    public String getNumTvaIntracom() {
        return numTVAIntracom;
    }

    public Client setAdresse(Adresse adresse) {
        this.adresse = adresse;
        return this;
    }

    public String getNomCourt() {
        return nomCourt == null ? getNom() : nomCourt;
    }

    public Client setNomCourt(String nomCourt) {
        this.nomCourt = nomCourt;
        return this;
    }

    public Client setNumTVAIntracom(String num) {
        numTVAIntracom = num;
        return this;
    }
}
