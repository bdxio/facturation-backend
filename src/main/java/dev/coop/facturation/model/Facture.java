package dev.coop.facturation.model;

import lombok.ToString;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

/**
 *
 * @author lforet
 */
@ToString
public class Facture extends HasSocieteCodeKey {

    private Client client;
    private LocalDate date;
    private final List<Ligne> lignes;
    private final List<Ligne> ristournes;

    public Facture(SocieteCodeKey id, List<Ligne> lignes, List<Ligne> ristournes) {
        super(id);
        this.lignes = lignes;
        this.ristournes = ristournes;
    }

    public Facture(Societe societe, int codeValue) {
        super(societe, codeValue);
        this.lignes = new ArrayList<>();
        this.ristournes = new ArrayList<>();
    }

    @Override
    public void updateCode() {
        this.setCode(this.getCodeValue());
        for(Ligne l : lignes){
            l.updateArticleCode();
        }
        for(Ligne l : ristournes){
            l.updateArticleCode();
        }
    }

    @Override
    public CodeFormatter.Prefix getCodePrefix() {
        return CodeFormatter.Prefix.FA;
    }

    public Client getClient() {
        return client;
    }

    public Facture setClient(Client client) {
        this.client = client;
        return this;
    }

    public LocalDate getDate() {
        return date;
    }

    public LocalDate getDateEcheance() {
        return date.plusDays(getSociete().getDelaiPaiement());
    }
    
    public Facture setDate(LocalDate date) {
        this.date = date;
        return this;
    }

    public List<Ligne> getLignes() {
        return lignes;
    }

    public Facture addLigne(Ligne ligne) {
        lignes.add(ligne);
        return this;
    }

    public List<Ligne> getRistournes() {
        return ristournes;
    }

    public Facture addRistourne(Ligne ristourne) {
        ristournes.add(ristourne);
        return this;
    }

    public Montant getTotalHT() {
        Montant total = new Montant(BigDecimal.ZERO);
        for (Ligne ligne : lignes) {
            total = total.add(ligne.getMontantHT());
        }
        return total;
    }

    public Montant getTotalTtc() {
        Montant total = new Montant(BigDecimal.ZERO);
        for (Ligne ligne : lignes) {
            total = total.add(ligne.getMontantTtc());
        }
        return total;
    }

    public Montant getTotalTva() {
        Montant total = new Montant(BigDecimal.ZERO);
        for (Ligne ligne : lignes) {
            total = total.add(ligne.getMontantTva());
        }
        return total;
    }

}
