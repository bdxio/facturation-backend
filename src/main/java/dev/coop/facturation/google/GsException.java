package dev.coop.facturation.google;

import dev.coop.facturation.FacturationException;

/**
 *
 * @author lfo
 */
public class GsException extends FacturationException {

    public GsException() {
    }

    public GsException(String message) {
        super(message);
    }

    public GsException(String message, Throwable cause) {
        super(message, cause);
    }

    public GsException(Throwable cause) {
        super(cause);
    }

    
}
