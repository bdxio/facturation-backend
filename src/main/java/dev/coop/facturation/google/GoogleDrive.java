package dev.coop.facturation.google;

import com.google.api.client.googleapis.media.MediaHttpUploader;
import com.google.api.client.http.ByteArrayContent;
import com.google.api.services.drive.Drive;
import com.google.api.services.drive.model.File;
import com.google.api.services.drive.model.FileList;
import dev.coop.facturation.configuration.GoogleConfiguration;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import javax.annotation.PostConstruct;
import java.io.IOException;
import java.util.Collections;
import java.util.UUID;

/**
 *
 * @author lfo
 */
@Slf4j
@Service
public class GoogleDrive {

    @Autowired
    private GoogleConfiguration googleConfiguration;

    private Drive drive;

    @PostConstruct
    public void setup() {
        drive = new Drive.Builder(googleConfiguration.getTransport(), googleConfiguration.getJsonFactory(), googleConfiguration.getHttpRequestInitializer())
                .setApplicationName(GoogleConfiguration.APPLICATION_NAME)
                .build();

    }

    public void upload(String parentId, String title, String mimeType, byte[] bytes) throws IOException {
        File file = new File();
        file.setName(title);
        file.setParents(Collections.singletonList(parentId));
        file.setMimeType(mimeType);
        ByteArrayContent content = new ByteArrayContent(mimeType, bytes);
        Drive.Files.Create create = drive.files()
                .create(file, content)
                .setQuotaUser(UUID.randomUUID().toString())
                .setSupportsTeamDrives(true);
        MediaHttpUploader uploader = create.getMediaHttpUploader();
        uploader.setDirectUploadEnabled(true);

        create.execute();
    }

    public String getFile(String name, String parentId) throws IOException {
            String q = String.format("'%s' in parents and name = '%s' and trashed = false", parentId, name);
            FileList result = drive.files().list()
                    .setQ(q)
                    .setSpaces("drive")
                    .setFields("nextPageToken, files(id, name)")
                    .setQuotaUser(UUID.randomUUID().toString())
                    .setSupportsTeamDrives(true)
                    .setIncludeItemsFromAllDrives(true)
                    .execute();

            return result.getFiles().stream()
                    .map(File::getId)
                    .findFirst()
                    .orElse(null);
    }
    
    public void deleteFile(String id) throws IOException {
        File toTrash = new File();
        toTrash.setTrashed(true);

        drive.files().update(id, toTrash)
                .setQuotaUser(UUID.randomUUID().toString())
                .setSupportsTeamDrives(true)
                .execute();
    }
}
