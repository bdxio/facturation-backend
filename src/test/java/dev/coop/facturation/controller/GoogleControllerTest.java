package dev.coop.facturation.controller;

import dev.coop.facturation.Facturation;
import dev.coop.facturation.google.GoogleDrive;
import dev.coop.facturation.model.Facture;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;

import java.io.IOException;
import java.util.List;
import java.util.Map;

@RunWith(SpringJUnit4ClassRunner.class)
@SpringBootTest(classes = Facturation.class, webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
class GoogleControllerTest {

    public static final String WORKSHEET_ID = "1S6hdN75AYQqdk9cfC2Ot6dORWZu6u3NfwWzUiPylUdU";
    public static final String GOOGLE_DRIVE_FOLDERID = "1-Owr1AUhBm9E00NdhiK0z_7A06J6WAmY";
    public static final String BDXIO = "BDXIO";
    @Autowired
    private GoogleController googleController;

    @Mock
    private GoogleDrive driveMock;

    @BeforeEach
    public void beforeEach() throws IOException {
        //prevent all google drive write!
        Mockito.doNothing().when(driveMock).upload(Mockito.any(), Mockito.any(), Mockito.any(), Mockito.any());
        googleController.setGoogleDrive(driveMock);
    }

    @Test
    void getFacturesBySociety() {
        googleController.importSheet(WORKSHEET_ID);
        List<Facture> facturesBySociety = googleController.getFacturesBySociety(BDXIO);
        Approvals.verifyAll("facturesBySociety", facturesBySociety);
        for(Facture facture: facturesBySociety){
            Assertions.assertNotNull(facture.getCode());
        }
    }

    @Test
    void createFileToUpload() {
        googleController.importSheet(WORKSHEET_ID);
        List<Facture> facturesBySociety = googleController.getFacturesBySociety(BDXIO);
        Map<String, byte[]> fileToUpload = googleController.createFileToUpload(GOOGLE_DRIVE_FOLDERID, facturesBySociety);
        Approvals.verifyAll("fileToUpload", fileToUpload.keySet());
    }

}