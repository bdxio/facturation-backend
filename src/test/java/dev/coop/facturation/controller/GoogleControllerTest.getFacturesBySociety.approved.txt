facturesBySociety[0] = Facture(client=Client(nom=Citruce, nomCourt=Citruce, numTVAIntracom=null, adresse=Adresse(rue1=101 rue de Sèvres, rue2=Lot 1665, rue3=, codePostal=75272, ville=Paris Cedex 06, pays=null)), date=2020-05-07, lignes=[Ligne(article=Article(description=Sponsoring BDX I/O catégorie Impériale, montant=2 500,00 €, unite=null, tva=NONE), description=Sponsoring BDX I/O 2020 catégorie Impériale, montantUnitaire=2 500,00 €, unite=null, tva=NONE, quantite=1)], ristournes=[])
facturesBySociety[1] = Facture(client=Client(nom=Mangas Gambling Engineering, nomCourt=Mangas Gambling Engineering, numTVAIntracom=null, adresse=Adresse(rue1=117 Quai de Bacalan, rue2=, rue3=, codePostal=33300, ville=BORDEAUX, pays=null)), date=2020-05-11, lignes=[Ligne(article=Article(description=Sponsoring BDX I/O catégorie Balthazar, montant=5 000,00 €, unite=null, tva=NONE), description=Sponsoring BDX I/O 2020 catégorie Balthazar, montantUnitaire=5 000,00 €, unite=null, tva=NONE, quantite=1)], ristournes=[])
facturesBySociety[2] = Facture(client=Client(nom=APSIDE SA, nomCourt=APSIDE SA, numTVAIntracom=null, adresse=Adresse(rue1=Agence de Bordeaux, rue2=23, quai de Paludate, rue3=, codePostal=33800, ville=BORDEAUX, pays=null)), date=2020-05-11, lignes=[Ligne(article=Article(description=Sponsoring BDX I/O catégorie Balthazar, montant=5 000,00 €, unite=null, tva=NONE), description=Sponsoring BDX I/O 2020 catégorie Balthazar, montantUnitaire=5 000,00 €, unite=null, tva=NONE, quantite=1), Ligne(article=Article(description=Remise , montant=-500,00 €, unite=null, tva=NONE), description=Remise 10% APSIDE, montantUnitaire=-500,00 €, unite=null, tva=NONE, quantite=1)], ristournes=[])
facturesBySociety[3] = Facture(client=Client(nom=AT INTERNET, nomCourt=, numTVAIntracom=null, adresse=Adresse(rue1=85 Avenue John Fitzgerald Kennedy, rue2=, rue3=, codePostal=33700, ville=MERIGNAC, pays=null)), date=2020-05-11, lignes=[Ligne(article=Article(description=Sponsoring BDX I/O catégorie Magnum, montant=1 000,00 €, unite=null, tva=NONE), description=Sponsoring BDX I/O 2020 catégorie Magnum, montantUnitaire=1 000,00 €, unite=null, tva=NONE, quantite=1)], ristournes=[])
facturesBySociety[4] = Facture(client=Client(nom=COLIBRI SAS, nomCourt=COLIBRI SAS, numTVAIntracom=null, adresse=Adresse(rue1=52 rue Bayen, rue2=, rue3=, codePostal=75017, ville=PARIS, pays=null)), date=2020-05-15, lignes=[Ligne(article=Article(description=Sponsoring BDX I/O catégorie Balthazar, montant=5 000,00 €, unite=null, tva=NONE), description=Sponsoring BDX I/O 2020 catégorie Balthazar, montantUnitaire=5 000,00 €, unite=null, tva=NONE, quantite=1)], ristournes=[])
facturesBySociety[5] = Facture(client=Client(nom=Ippon Technologies, nomCourt=Ippon Technologies, numTVAIntracom=null, adresse=Adresse(rue1=43 avenue de la Grande Armée, rue2=, rue3=, codePostal=75116, ville=Paris, pays=null)), date=2020-05-23, lignes=[Ligne(article=Article(description=Sponsoring BDX I/O catégorie Impériale, montant=2 500,00 €, unite=null, tva=NONE), description=Sponsoring BDX I/O 2020 catégorie Impériale, montantUnitaire=2 500,00 €, unite=null, tva=NONE, quantite=1), Ligne(article=Article(description=Sponsoring BDX I/O, montant=500,00 €, unite=null, tva=NONE), description=Sponsoring BDX I/O, montantUnitaire=500,00 €, unite=null, tva=NONE, quantite=1)], ristournes=[])
facturesBySociety[6] = Facture(client=Client(nom=Clever Cloud, nomCourt=Clever Cloud, numTVAIntracom=null, adresse=Adresse(rue1=3 rue de l'Allier, rue2=, rue3=, codePostal=44000, ville=Nantes, pays=null)), date=2020-05-20, lignes=[Ligne(article=Article(description=Sponsoring BDX I/O catégorie Impériale, montant=2 500,00 €, unite=null, tva=NONE), description=Sponsoring BDX I/O 2020 catégorie Impériale, montantUnitaire=2 500,00 €, unite=null, tva=NONE, quantite=1)], ristournes=[])
facturesBySociety[7] = Facture(client=Client(nom=AUDENSIEL TECHNOLOGIES, nomCourt=AUDENSIEL TECHNOLOGIES, numTVAIntracom=null, adresse=Adresse(rue1=5 rue du Château trompette, rue2=, rue3=, codePostal=33000, ville=Bordeaux, pays=null)), date=2020-05-26, lignes=[Ligne(article=Article(description=Sponsoring BDX I/O catégorie Impériale, montant=2 500,00 €, unite=null, tva=NONE), description=Sponsoring BDX I/O 2020 catégorie Impériale, montantUnitaire=2 500,00 €, unite=null, tva=NONE, quantite=1)], ristournes=[])
